pkg/util/hostname.go

```diff
@@ -121,54 +155,192 @@ func saveHostnameData(cacheHostnameKey string, hostname string, provider string)
 	return hostnameData
 }
 
+// ResolveSourcesWithState will merge two HostnameMap sources, providing the expected
+// resolved output.
+func ResolveSourcesWithState(newSources, stateSources HostnameMap) (HostnameMap, bool) {
+	resolvedSources := HostnameMap{}
+	stateChange := false
+
+	if stateSources == nil {
+		return newSources, true
+	}
+
+	for _, stage := range resolutionPipeline {
+		log.Debug("Getting hostname collected by: %s", stage.provider)
+
+		newH, newOk := newSources[stage.provider]
+		stateH, stateOk := stateSources[stage.provider]
+
+		if !newOk && !stateOk {
+			// missing source
+			continue
+		}
+
+		h := newH
+		if stage.reliable {
+			if newOk {
+				if newH != stateH {
+					stateChange = true
+				}
+			} else {
+				if stateOk {
+					stateChange = true
+					log.Warnf("Reliable source %s no longer configured", stage.provider)
+				}
+
+				// reliable source did not resolve thus it does not apply
+				continue
+			}
+		} else {
+			if newOk {
+				if newH != stateH {
+					stateChange = true
+				}
+			} else {
+				if stateOk {
+					// here we should always use stored state - an unreliable source failed
+					h = stateH
+					log.Info("an unreliable source %s did not resolve a hostname as expected, using stored state: %s",
+						stage.provider, stateH)
+				} else {
+					// resolution stage did not resolve as expected
+					continue
+				}
+			}
+		}
+
+		resolvedSources[stage.provider] = h
+	}
+
+	return resolvedSources, stateChange
+}
+
 // GetHostnameData retrieves the host name for the Agent and hostname provider, trying to query these
 // environments/api, in order:
+// * Fargate
 // * GCE
 // * Docker
 // * kubernetes
 // * os
 // * EC2
 func GetHostnameData() (HostnameData, error) {
+
 	cacheHostnameKey := cache.BuildAgentKey("hostname")
 	if cacheHostname, found := cache.Cache.Get(cacheHostnameKey); found {
 		return cacheHostname.(HostnameData), nil
 	}
 
 	var hostName string
-	var err error
+	var fqdn string
 	var provider string
+	var err error
+
+	live, err := liveSourcer()
+	if err != nil {
+		return HostnameData{}, err
+	}
+
+	state, err := stateSourcer()
+	if err != nil {
+		log.Warn("stateful hostname unavailable, will rely on live sources")
+	}
+
+	sources, stateChange := configSourceResolver(live, state)
+	for _, stage := range resolutionPipeline {
+		log.Debug("Getting hostname collected by: %s", stage.provider)
+
+		if h, ok := sources[stage.provider]; ok {
+
+			if h != "" || stage.provider == "fargate" {
```

Comment:
for `fargate` we called `hostnameData := saveHostnameData(cacheHostnameKey, "", "")` whereas we will call `hostnameData := saveHostnameData(cacheHostnameKey, "fargate", "")`. I do not know if it is an issue.